sycl/source/detail/graph_impl.hpp

```diff
@@ -344,6 +354,32 @@ class node_impl {
     return NumberOfNodes;
   }
 
+  /// Duplicate recursively a node and its successors
+  /// @param NodesMaps map that associates old nodes pointer to their
+  /// duplicates. This map is populated by this function.
+  /// @return a shared_ptr to the new duplicated node
+  std::shared_ptr<node_impl> duplicateNodeAndSuccessors(
+      std::map<node_impl *, std::shared_ptr<node_impl>> &NodesMaps) {
+    // If the node has already been copied, we skip it
+    if (NodesMaps.find(this) != NodesMaps.end())
+      return NodesMaps[this];
+
+    std::shared_ptr<node_impl> CpyNode;
+    if (MCGType == sycl::detail::CG::None) {
+      CpyNode = std::make_shared<node_impl>();
+      CpyNode->MCGType = sycl::detail::CG::None;
+    } else {
+      CpyNode = std::make_shared<node_impl>(MCGType, getCGCopy());
+    }
+    NodesMaps.insert({this, CpyNode});
+    for (auto &NextNode : MSuccessors) {
+      auto Successor = NextNode->duplicateNodeAndSuccessors(NodesMaps);
+
+      CpyNode->registerSuccessor(Successor, CpyNode);
+    }
+    return CpyNode;
+  }
+
```

Comment:
depth first search does not allow us to register successor nodes correctly as it is.
I found a way to duplicate subgraph using another method (based on the list of the scheduled nodes).
However we still need duplicateNodeAndSuccessors method to duplicate graph_impl because nodes are not scheduled in graph_impl.
That said graph_impl duplication is not used for the moment. So if you prefer I can remove all this support.

Reply:
I finally removed graph_impl copy constructor and duplicateNodeAndSuccessors function